src/Common/QueryProfiler.cpp

```diff
@@ -40,8 +41,19 @@ namespace
     /// to ignore delivered signals after timer_delete().
     thread_local bool signal_handler_disarmed = true;
 
+    /// Don't permit too many threads be busy inside profiler,
+    /// which could slow down the system in some environments.
+    std::atomic<Int64> concurrent_invocations = 0;
+
     void writeTraceInfo(TraceType trace_type, int /* sig */, siginfo_t * info, void * context)
     {
+        SCOPE_EXIT({ concurrent_invocations.fetch_sub(1, std::memory_order_relaxed); });
+        if (concurrent_invocations.fetch_add(1, std::memory_order_relaxed) > 100)
```

Comment:
100 seems to be a random constant. I think it should be somehow related to the number of processors. For example, just replace it with the number of cores in the system. Because sampling more than 2 threads on a single core, while all other CPUs are also busy with sampling at the same time looks too much imho.

Reply:
Yes, also, it was not certain for me. But it is not like the number of cores. For example, if the system has a large number of cores, like 256, I still don't expect that more than 100 threads will be inside the profiler at the same time. If we think the profiler should take not more than 10% (but I'd aim for less than 1%), then the threshold should be 25 or less. So, 100 is something like a high enough safety threshold, regardless of the number of cores.

Finally, the profiler degrades when it is a real-time profiler for a lot of idle threads (thousands).

Reply:
Okay let's keep 100